jax/experimental/jax2tf/jax2tf.py

```diff
@@ -126,6 +126,122 @@ def add_unit(v: TfValOrUnit, aval: core.AbstractValue):
+
+def categorize(prim: core.Primitive, *args: TfVal, **kwargs) \
+    -> List[Limitation]:
+  """
+  Given a primitive and a set of parameters one would like to pass to it,
+  categorize identifies the potential limitations the call would encounter when
+  converted to TF through jax2tf.
+
+  Args:
+    prim: the primitive to call.
+    args: the arguments to pass to prim.
+    kwargs: the keyword arguments to pass to prim.
+
+  Returns:
+    A list of limitations
+  """
+  limitations: List[Limitation] = []
+  all_devices = ["CPU", "GPU", "TPU"]
+
+  def _report_failure(error_type: str, msg: str,
+                      devs: Sequence[str] = all_devices) -> None:
+    limitations.append(Limitation(prim.name, error_type, msg, tuple(devs)))
+
+  tf_unimpl = functools.partial(_report_failure, "Missing TF support")
+
+  def _to_np_dtype(dtype):
+    try:
+      dtype = to_jax_dtype(dtype)
+    except:
+      pass
+    return np.dtype(dtype)
+
+  if prim in [lax.min_p, lax.max_p]:
+    np_dtype = _to_np_dtype(args[0].dtype)
+    if np_dtype in [np.bool_, np.int8, np.uint16, np.uint32, np.uint64,
+                    np.complex64, np.complex128]:
+      tf_unimpl(f"{prim.name} is unimplemented for dtype {np_dtype}")
+
+  return limitations
+
+def prettify(limitations: Sequence[Limitation]) -> str:
+  """Constructs a summary .md file based on a list of limitations."""
+  limitations = sorted(list(set(limitations)))
+
+  def _pipewrap(columns):
+    return '| ' + ' | '.join(columns) + ' |'
+
+  title = '# Primitives with limited support'
+
+  column_names = [ 'Affected primitive'
+                 , 'Type of limitation'
+                 , 'Description'
+                 , 'Devices affected' ]
+
+  table = [column_names, ['---'] * len(column_names)]
+
+  for lim in limitations:
+    table.append([ lim[0], lim[1], lim[2], ', '.join(lim[3]) ])
+
+  return title + '\n\n' + '\n'.join(line for line in map(_pipewrap, table))
+
+@SingletonDecorator
+class Categorizer:
+  """Conveniently wraps the categorization utils of jax2tf in a Singleton.
+
+  Args:
+    inner_categorizer: the categorization function to call when examining
+      primitives (default: categorize).
+    inner_prettifier: the prettifying function to call at the exit of the
+      program to build a summary of the encountered limitations
+      (default: prettify).
+  """
+  def __init__(self, inner_categorizer: Callable = categorize,
+               inner_prettifier: Callable = prettify):
+    self.all_limitations: Sequence[Limitation] = []
+    self.inner_categorizer = inner_categorizer
+    self.inner_prettifier = inner_prettifier
+
+  def wrap(self, prim: core.Primitive, func: Callable) -> Callable:
+    """
+    Wraps a primitive and its corresponding TF implementation with the inner
+    categorizer.
+    """
+    def wrapper(*args, **kwargs):
+      limitations = self.inner_categorizer(prim, *args, **kwargs)
+      self.all_limitations += limitations
```

Comment:
@gnecula here is where the limitations are added to the state within the singleton